python/rmm/_lib/device_pointer.pyx

```diff
@@ -0,0 +1,13 @@
+from rmm._lib.lib cimport c_free, cudaStream_t
+
+cdef class DevicePointer:
+    def __cinit__(self, ptr, stream=0):
+        self.c_ptr = <void*><size_t>(ptr)
+        self.c_stream = <cudaStream_t><size_t>(stream)
+
+    def __init__(self, ptr, stream=0):
+        self.ptr = ptr
+        self.stream = stream
+
+    def __dealloc__(self):
+        c_free(self.c_ptr, self.c_stream)
```

Comment:
Am a little wary of using `__dealloc__` to handle cleanup as this is relying on the Python runtime to run this. Would it be possible to push this into a C++ deconstructor or object?

Reply:
As one idea, we might use a `std::unique_ptr` with a custom deleter to hold the pointer and later free it.

Reply:
I like that better as well.

Reply:
`__dealloc__` works fine in my experience and freeing memory is its intended and advertised purpose. It is `__del__` that cannot be treated as a destructor and that Python makes no guarantees about.

Reply:
Unfortunately I have not had the same experience. Tried to use this for RAII previously and had issues where it was not run. Usually ran into issues when the process exits in some unclean way.

Reply:
I don't think that's a big deal for us as if free isn't called at program exit the driver will clean up the memory allocations. That being said we should look at the generated cpp for what it generates for the `__dealloc__` here as in my experience this typically runs after the Python interpreter is gone so it isn't reliant on Python.

Reply:
`__dealloc__` proceeds through `tp_dealloc`. Relevant snippet from Cython docs below:

> This is done using the `tp_dealloc` slot, which can be customized in Cython by implementing `__dealloc__`.

ref: https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#controlling-deallocation-and-garbage-collection-in-cpython

This is run by the Python interpreter when the reference count goes to zero.

> The destructor function is called by the `Py_DECREF()` and `Py_XDECREF()` macros when the new reference count is zero.

ref: https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc

Reply:
I looked briefly into using `unique_ptr` with a custom deleter as you suggested @jakirkham, but ran into something similar to https://stackoverflow.com/questions/53582945/wrapping-c-code-with-function-pointer-as-template-parameter-in-cython

Basically Cython expects template arguments to be types. I tried to experiment with the hack mentioned in the solution but couldn't get it working.

Any ideas? I'd be happy to dig further if you feel strongly about not using `__dealloc__`, but my view is we leave it as is and revisit if GC is problematic.

Reply:
Yep that makes sense. Was able to do something similar with a C++ class (Slacked details).

Happy to defer to your judgement here about the best path forward. 🙂

Reply:
[These test]( https://github.com/cython/cython/blob/e2d93e20065607f2941f875ef29f077ba511d83b/tests/run/cpp_smart_ptr.pyx#L42 ) may be useful for future exploration.